### ISSUE TITLE
Clarification on the data generate code

### ISSUE THREAD

**Issue by padmaksha18** [yuke93/RL-Restore#5]
Hi Yue, Many thanks for sharing this repo, very useful to us!! I was trying to run the same code with octave but getting some error in this line
curr_dat_sz = store2hdf5(savepath, batchdata, batchlabs, ~created_flag, startloc, chunksz);

I was trying to use the save function in octave with options as hdf5 but it is throwing some error with the arguments. Can you please help me with how to implement the same functionality in octave.

Secondly, can anyone please explain to me what is the logic applied and those calculations done before storing the data in hdf5 format. I really did not get why it is divided by 64 and all. Kindly help me here.

Many thanks.

**Comment by yuke93** [comment]
Hi @padmaksha18, 

Thanks for your question. The `chunksz` changes the way how data are stored in an hdf5 file (refer to this [link](https://support.hdfgroup.org/HDF5/doc/Advanced/Chunking/index.html)). There is performance gain only when you need to manipulate a subset of data. In our case, `chunksz` is not necessary since we directly read a whole hdf5 in the memory. I am not familiar with octave, but please check if you can save it correctly by ignoring the argument of `chunksz`.

**Comment by padmaksha18** [comment]
hi @yuke93 ,
Many thanks for your response. But even if I donot consider the chunkz argument, the data is stored in a struct in this line
startloc = struct('dat',[1,1,1,totalct+1], 'lab', [1,1,1,totalct+1]); 

and here, the totalct value is increased by 1 for each iteration. Since the number of times the loop will execute depends on the chunkz value, how am i supposed to handle this without the chunkz value. Kindly help me here in understanding how the loop is executing. After that, i think i just need to save the 
struct into hdf5 format. Kindly advise.

Thanks

**Comment by padmaksha18** [comment]
hi @yuke93 ,

I just need an explanation of the matlab data generation code , the logic behind each step . I want to convert it to Python code. Can you please help me understand it or some reference to it. I cant find anything related in the research paper.

Thanks

**Comment by yuke93** [comment]
Hi @padmaksha18,

You may modify the code in the following way and see whether it works.

1. In generate_train.py, change the last loop by these 2 lines:
`startloc = struct('dat',[1,1,1,1], 'lab', [1,1,1,1]);`
`store2hdf5(savepath, data, label, ~created_flag, startloc);`

2. In store2hdf5.m, replace line 31 and 32 with:
`h5create(filename, '/data', [dat_dims(1:end-1) length(data)], 'Datatype', 'single');`
`h5create(filename, '/label', [lab_dims(1:end-1) length(labels)], 'Datatype', 'single');`

In this way, you do not need "chunksz" anymore.

**Comment by yuke93** [comment]
@padmaksha18 

It seems the loop makes confusion. In the loop, the data are written into an hdf5 file, one chunk after another (the total length is flexible). By the above modification, all data are written into an hdf5 file at one time (with total length fixed). You may use the `h5py` library in python.

**Comment by padmaksha18** [comment]
hi @yuke93 ,
Many thanks for all your help. It works that way perfectly without the chunkz argument. Actually, I cannot procure a matlab license at the moment for my project and therefore was considering to convert the entire data generation code into python. As far as i understood, i need to convert 3 files , the data generate code, the store hdf5 code and the noise combination code. I was just asking if there is any wiki etc where the logic or step by step algorithm for each of them is explained , which will help me to understand them overall. Otherwise I will have to do a line by line conversion. If you could help me here, that would be really great for me. I hope I am not asking too much. Many thanks again!

**Comment by yuke93** [comment]
Hi @padmaksha18 ,

I can give you a brief instruction. 

1. For store2hdf5.m, you can use the `h5py` library in Python. There is a simple example [here](http://docs.h5py.org/en/stable/quick.html#appendix-creating-a-file).

2. For generate_train.m, it adds mixed distortions to each image and crops the image into small patches. You may use `opencv-python` or `Pillow` library to add these distortions, and the cropping operation should be easy to implement.

3. For noise_combination.m, it calculates all possible combinations of mixed distortions within a certain level. The level of mixed distortion is illustrated in our [supplementary material](http://mmlab.ie.cuhk.edu.hk/projects/RL-Restore/support/supp.pdf) (Section 2, "Details of Test Sets. ").

**Comment by padmaksha18** [comment]
HI @yuke93 ,

Many thanks for providing me this overview. It is much clear to me now. Just one question I had regarding that data and label that you are segregating from the data. In the storehdf5 file, you have mentioned that 
 % *data* is C*H*W*N matrix of images should be normalized (e.g. to lie between 0 and 1) beforehand
  % *label* is D*N matrix of labels (D labels per sample) 

I am not clear about the label part, like what did you mean by D labels per sample.

Again, this part of the code is written for this purpose only , if I am not wrong.

for x = 1 : stride : hei-size_input+1
                for y = 1 :stride : wid-size_input+1
                    subim_input = im_input(x : x+size_input-1, y : y+size_input-1, :);
                    subim_label = im_label(x+padding : x+padding+size_label-1, y+padding : y+padding+size_label-1, :);
                    count=count+1;
                    data(:, :, :, count) = subim_input;
                    label(:, :, :, count) = subim_label;
                end

Can you please shortly explain what it is overall doing. Will be great for me. 

thanks,
Padmaksha

**Comment by yuke93** [comment]
Hi @padmaksha18 ,

I am sorry that the annotation is misleading. The "label" should be also a CHWN matrix of images. I have corrected this annotation. Hope it addresses the confusion.

**Comment by padmaksha18** [comment]
Hi @yuke93 ,
Many thanks for correcting it, it was causing lot of confusions :) . Can you please explain me the algorithm behind this loop, like what you are trying to do in these steps. I know i am consuming your time a lot, but i dont find anyone else in this repo to ask for . Please advise.

for x = 1 : stride : hei-size_input+1
for y = 1 :stride : wid-size_input+1
subim_input = im_input(x : x+size_input-1, y : y+size_input-1, :);
subim_label = im_label(x+padding : x+padding+size_label-1, y+padding : y+padding+size_label-1, :);
count=count+1;
data(:, :, :, count) = subim_input;
label(:, :, :, count) = subim_label;
end

Many thanks.
Padmaksha

**Comment by yuke93** [comment]
Hi @padmaksha18 ,

In this loop, I am trying to crop training pairs of distorted and ground-truth patches. The `im_input` and `im_label` are distorted and ground-truth images, respectively. The `subim_input` and `subim_label` are cropped patches, and I put them into array `data` and `label`.

I only consider color images. For grayscale images, I just do "continue" to ignore them.

**Comment by padmaksha18** [comment]
hi @yuke93 ,
Thanks a lot for the response. I did not get what you meant here by "patches" , do you mean a portion of the images or the entire image as a whole.

I have some other doubts as well, please help me in them.

1. I did not understand what is the N here. Since you are introducing all 3 types of distortions step by step on every image, so why is then N required here,

2. I could not convert this line to python.
`
level_temp = level_vec(ceil(rand(1)*level_num), :);`

I tried something like this,
`level_temp = level_vec(math.ceil(np.random.rand(1,1), level_num)), range())`

I dont know what to put in the range() argument for the range. Please help here.

3. I got some codes for Gaussian noise insertion, can i use them as is here.
https://stackoverflow.com/questions/17190649/how-to-obtain-a-gaussian-filter-in-python

4.  I did not understand what is the count dimension here in this line.
```

data(:, :, :, count) = subim_input; 
 label(:, :, :, count) = subim_label;
```

And again on what basis you are assigning it to order.

`order = randperm(count);`

Please help me here.

Kindly help me with these queries.

Thanks a lot

**Comment by yuke93** [comment]
Hi @padmaksha18 ,

I am afraid I cannot answer every question for you at the moment. It seems there will be a lot of questions before you can reimplement the codes in Python. I will see if I can write an example for you after the ICCV deadline. Thanks for your understanding.

**Comment by padmaksha18** [comment]
HI @yuke93 ,
Many thanks for all your help till now. Surely will wait till your conference. All the best for that!!

If you get some time in between, can you please put some comments on this code. At least I will try to progress on the basis of that till you are free. 

Many thanks again!
Padmaksha

**Comment by yuke93** [comment]
Hi @padmaksha18 ,

About your last few questions:

1. You may set N>1 if you want to generate more diverse data. The larger N is, the more data are generated.

2. May try `level_temp = level_vec[np.ceil(np.random.rand() * level_num), :]`.

3. Just try it, and observe the generated image to check.

4. The "count" represents the total number of training pairs in the dataset. The "order" is used for random shuffle.

**Comment by padmaksha18** [comment]
HI Yuke,
Extremely sorry for coming back so late on this. I will try this changes and come back to you asap.

Many thanks
Padmaksha

**Comment by padmaksha18** [comment]
hi @yuke93 ,

This is me again :) . I have started converting the code and need your help here badly.
```
startloc = struct('dat',[1,1,1,1], 'lab', [1,1,1,1]);
store2hdf5(savepath, data, label, ~created_flag, startloc);
```

In the above code, what is the significance of the 'startloc' variable. I did not at all understand the calculations that you are doing with this value in the storehdf5 file. Stackoverflow says dict is an equivalent of Matlab struct datatype but I am not getting any clue of how it is a dict and how you have used it later on in storehdf5. Kindly explain me.

```
h5create(filename, '/data', [dat_dims(1:end-1) length(data)], 'Datatype', 'single');
h5create(filename, '/label', [lab_dims(1:end-1) length(labels)], 'Datatype', 'single');
```

In this part, I am writing it like this in python,

     ```
       import h5py
        with h5py.File(filename, "w") as f:
            f.create_dataset("data", (100, 100), dtype='i')
            f.create_dataset("label", (100, 100), dtype='i')
```

Is it a correct implementation in Python, please confirm. Again, the datatype is int here and in Matlab code it is single.

**Comment by padmaksha18** [comment]
hi @yuke93 ,

This is me again :) . I have started converting the code and need your help here badly.
```
startloc = struct('dat',[1,1,1,1], 'lab', [1,1,1,1]);
store2hdf5(savepath, data, label, ~created_flag, startloc);
```

In the above code, what is the significance of the 'startloc' variable. I did not at all understand the calculations that you are doing with this value in the storehdf5 file. Stackoverflow says dict is an equivalent of Matlab struct datatype but I am not getting any clue of how it is a dict and how you have used it later on in storehdf5. Kindly explain me.

```
h5create(filename, '/data', [dat_dims(1:end-1) length(data)], 'Datatype', 'single');
h5create(filename, '/label', [lab_dims(1:end-1) length(labels)], 'Datatype', 'single');
```

In this part, I am writing it like this in python,

     ```
       import h5py
        with h5py.File(filename, "w") as f:
            f.create_dataset("data", (100, 100), dtype='i')
            f.create_dataset("label", (100, 100), dtype='i')
```

Is it a correct implementation in Python, please confirm. Again, the datatype is int here and in Matlab code it is single. Will it cause any issue?

Many thanks!!

**Comment by yuke93** [comment]
Yes, go ahead. You may use int (range 0-255) or float (range 0.0-1.0).

**Comment by padmaksha18** [comment]
Thanks @yuke93 . But my main doubt was with the 'startloc' variable. I am not getting how it is used in the
storehdf5.m code. Can you please explain me once. Thanks again!

**Comment by padmaksha18** [comment]
HI @yuke93 ,

Please confirm, if I have converted these two lines correctly. The dimensions especially.
```
h5create(filename, '/data', [dat_dims(1:end-1) length(data)], 'Datatype', 'single');
h5create(filename, '/label', [lab_dims(1:end-1) length(labels)], 'Datatype', 'single');
```

Python

 ```
with h5py.File(filename, "w") as f:
            f.create_dataset("data", (dat_dims[: -1], len(data)), dtype='i')
            f.create_dataset("label", (lab_dims[: -1], len(labels)), dtype='i')
```
Please confirm me if the dimensions are correct. 
As far as i got, you are creating the h5 files with dimensions of 
size lab_dims(everything except last element) *  len(data). 

Please correct me here. Thanks

**Comment by padmaksha18** [comment]
Hi @yuke93 ,

Any chance you are free now. I have lots of doubts which I want to get clarified one by one. Many thanks in advance.

Padmaksha

**Comment by padmaksha18** [comment]
Hi Yuke,

Hope you are doing well.

I have used your github code to restore some document images after adding
the three types of noises as mentioned in your code.

I have attached some sample of prediction results that I got. The left hand
side images are from the  training set images only and I have added some
noise to them  and used them as test images to the models generated from my
training to restore them back to original. But apart from these few in most
of  the images that I tried the images are getting more blurred than the
input blurred image itself . The results are very poor actually.

While observing the tensorboard I found the following pattern - the reward
function was increasing initially but gradually got saturated at a point.

Can you please help me with what can be the possible reason for this.

Training stats:
1. I trained with 250 document images and it trained for over 10 hours
before getting saturated.

[image: image.png]

[image: image.png]









 Besides that, I have some more doubts regarding the code conversion to
python. I have done most of it but need to clarify at some steps.

Thanks
Padmaksha

**Comment by Bander-Almalki** [comment]
Hi Yuke! 
I run the code and every thing was fine until I reached the following command 
`data/train/generate_train.m`
  the file opened on Matlab, but the following error appears when I try to run the code on Matlab. 
`>> h5info(validation, location)
Undefined function or variable 'validation'.`
 
Any suggestions?

**Comment by yuke93** [comment]
Hi @Bander-Almalki,

Did you change the code? This error occurred because you did not define the variable `validation`, the path of an hdf5 file, before you use this variable.

**Comment by Bander-Almalki** [comment]
> 
> 
> Hi @Bander-Almalki,
> 
> Did you change the code? This error occurred because you did not define the variable `validation`, the path of an hdf5 file, before you used this variable.

Hi Yuke! thanks for replying. I figured out the problem. it was because I didn't copy the DIV2K folder into /data/train directory. After doing so, it is working properly.

**Comment by yuke93** [comment]
> > Hi @Bander-Almalki,
> > Did you change the code? This error occurred because you did not define the variable `validation`, the path of an hdf5 file, before you used this variable.
> 
> Hi Yuke! thanks for replying. I figured out the problem. it was because I didn't copy the DIV2K folder into /data/train directory. After doing so, it is working properly.

Great!
